Add support for commands/shortcodes sentence-capitalizing long form

psomhorst opened the pull request:
In some languages, the long form is always capitalized. In other languages long forms are not capitalized, which makes it harder to use acronyms at the start of sentences. To prevent having to use `{{< acr KEY case=sentence >}}` throughout a document, this PR introduces capitalized versions of the `\acr` and `\acrs` command, as well as the `{{< acr ... >}}` shortcode. 

When using `\Acr`, `\Acrs`, `{{< Acronym ... >}}` or `{{< Acr ... >}}`, `case` is set to `'sentence'` and `case_target` is set to `'long'`. 

This last part might be up for discussion: by setting `case_target`, the user is prevented to set it to `'both'` or `'short'`. Maybe the code should be expanded to 'upgrade' `case_target` to `'both'` if the user sets it to `'short'`?

Commit:
Add support for commands/shortcodes sentence-capitalizing long form of acronyms

rchaput commented:
Excellent idea! I agree that `case=sentence`, `case=upper`, etc., are a bit long. When used only once or twice, it's OK, but when using throughout a document, it can be cumbersome.

I think it would be helpful to also have `\ACR`, `\ACRS`, `{{< ACR >}}` and `{{< ACRS >}}`. I'll try to add this to your PR.

However, I don't understand why changing the `case_target` option in these cases? Why setting it to `long`?
Maybe we could set it to `both` if it's not set, but use the given value if it's set.

Also, I'm not sure we need to change the example (the goal is not to put every possible feature in the example); but we need to have tests that both demonstrate how this feature works, and ensures it actually work. I'll also try to add it when I have time.

psomhorst commented:
> However, I don't understand why changing the `case_target` option in these cases? Why setting it to `long`? Maybe we could set it to `both` if it's not set, but use the given value if it's set.

My goal is to **only** capitalize the long form. If you would do `{{< Acr key case_target=short >}}` it would **not** capitalize the long form, but **would** capitalize the short form. I don't think that is the behavior I would expect. However, since long is the default case target, only adapting it specifically would change the behavior, so I don't think it's necessary?
 
> Also, I'm not sure we need to change the example (the goal is not to put every possible feature in the example); but we need to have tests that both demonstrate how this feature works, and ensures it actually work. I'll also try to add it when I have time.

It took me *quite* a while to even discover this was possible. Examples would be *very* helpful for people to discover all possibilities.

Commit:
Add helper function to support simplified shortcodes

In order to support more shortcodes that simplify usage (such as "Acr",
"Acrs", etc.), added a new function that takes default values and return
a shortcode function.

Commit:
Add more simplified shortcodes

In the same idea as "Acr" and "Acronym", added "shortcuts" shortcodes:
- plural;
- sentence case + plural;
- upper case;
- upper case + plural.

Commit:
Update filter to also support simplified commands

Same as for the shortcodes, the filter now supports:
- ACR (upper case);
- ACRS (upper case + plural).

Note that, because of how I implemented the regex, an user could use
a bizarre command such as `AcR` (or any variation). Technically, this
would count in the `else` part, as if it was `acr`.

I'm not sure whether it would hinder the performance to ensure that only
`acr`, `Acr`, or `ACR` can be used. Because the user should not try this
anyway, I don't think it would be worth it.

                -- Technically, this `else` is also triggered when the command
                -- is `AcR` (or any variation like that). I'm not sure whether
                -- it would be better, in terms of performance, to ensure that
                -- only `acr`, `Acr`, or `ACR` can be used.

rchaput commented:
> My goal is to **only** capitalize the long form. If you would do `{{< Acr key case_target=short >}}` it would **not** capitalize the long form, but **would** capitalize the short form. I don't think that is the behavior I would expect. However, since long is the default case target, only adapting it specifically would change the behavior, so I don't think it's necessary?

I am mainly interested in providing sane defaults for most users; and I am still not sure what the "most used case" is. (I need to think a bit more about it)

> It took me _quite_ a while to even discover this was possible. Examples would be _very_ helpful for people to discover all possibilities.

I see your point. I tried to document the features, but maybe it gets "lost" in the documentation. A more complete example file that links to the docs for details could be a nice addition.

In the meantime, I added a few commits to improve your PR; the important change is the addition of `acrs`, `acronyms`, `Acrs`, `Acronyms`, `ACR`, `ACRS` (inspired by your PR #51).

agerada commented:
I think the proposed changes here are useful. In retrospect, perhaps a more flexible design choice with case modifications would have been to just separate out the case by short and long, e.g., `\acr[case_short=upper, case_long=sentence]{..}`. This is more verbose but gives full flexibility (all permutations are available to the user). Maybe something to consider in a future version, as the verbosity issue would be sorted by this PR.

I still think most users are highly likely to only modify the case for long names in practice. 

If the documentation is clear that `\Acr` only modifies the long name, then I think the original proposal (if `case_target` = `short`, then internally change to `both`) is reasonable.

rchaput commented:
> I think the proposed changes here are useful. In retrospect, perhaps a more flexible design choice with case modifications would have been to just separate out the case by short and long, e.g., `\acr[case_short=upper, case_long=sentence]{..}`. This is more verbose but gives full flexibility (all permutations are available to the user). Maybe something to consider in a future version, as the verbosity issue would be sorted by this PR.

I like this idea. Let's keep it in mind for a future version; but we should have a simplified option as well. For example, setting `\acr[case_short_upper, case_long=sentence]` but also keeping `\acr[case=upper]` if we do not want/need that much flexibility...

> I still think most users are highly likely to only modify the case for long names in practice.

I don't have a strong opinion on this, so I'll trust the both of you

> If the documentation is clear that `\Acr` only modifies the long name, then I think the original proposal (if `case_target` = `short`, then internally change to `both`) is reasonable.

What if we set `case_target` to `long` only when unset? Thus, users that set `case_target=short` would still get the exact behavior described by this option, and if they want both, they set `case_target=both` (which is the same number of characters, so not a verbosity problem...)

If you are right that most users will want to change the long name, then the "default" case is `\Acr` (without changing `case_target`), which is perfect.
